aiida/tools/groups/grouppaths.py

```diff
@@ -0,0 +1,321 @@
+        """Compare less-than operator of path and type string to another ``GroupPath`` object."""
+        if not isinstance(other, GroupPath):
+            return NotImplemented
+        return (self.path, self.type_string) < (other.path, other.type_string)
+
+    @property
+    def path(self):
+        # type: () -> str
+        """Return the path string."""
+        return self._path_string
+
+    @property
+    def path_list(self):
+        # type: () -> List[str]
+        """Return a list of the path components."""
+        return self._path_list[:]
+
+    @property
+    def key(self):
+        # type: () -> str
+        """Return the final component of the the path."""
+        if self._path_list:
+            return self._path_list[-1]
+        return None
+
+    @property
+    def delimiter(self):
+        # type: () -> str
+        """Return the delimiter used to split path into components."""
+        return self._delimiter
+
+    @property
+    def type_string(self):
+        # type: () -> str
+        """Return the type_string used to query for and instantiate a ``Group`` with."""
+        return self._type_string
+
+    @property
+    def parent(self):
+        # type: () -> Optional[GroupPath]
+        """Return the parent path."""
+        if self.path_list:
+            return GroupPath(
+                self.delimiter.join(self.path_list[:-1]),
+                type_string=self.type_string,
+                warn_invalid_child=self._warn_invalid_child
+            )
+        return None
+
+    def __truediv__(self, path):
+        # type: (str) -> GroupPath
+        """Return a child ``GroupPath``, with a new path formed by appending ``path`` to the current path."""
+        if not isinstance(path, str):
+            raise TypeError('path is not a string: {}'.format(path))
+        path = self._validate_path(path)
+        child = GroupPath(
+            path=self.path + self.delimiter + path if self.path else path,
+            type_string=self.type_string,
+            warn_invalid_child=self._warn_invalid_child
+        )
+        return child
+
+    def __getitem__(self, path):
+        # type: (str) -> GroupPath
+        """Return a child ``GroupPath``, with a new path formed by appending ``path`` to the current path."""
+        return self.__truediv__(path)
+
+    @property
+    def group(self):
+        # type: () -> Optional[orm.Group]
+        """Return the concrete group associated with this path."""
+        try:
+            if self.type_string is not None:
+                return orm.Group.objects.get(label=self.path)
+            return orm.Group.objects.get(label=self.path, type_string=self.type_string)
+        except NotExistent:
+            return None
+
+    @property
+    def is_virtual(self):
+        # type: () -> bool
+        """Return whether there is a concrete group associated with this path."""
+        # Note we don't test (self.group is None) here, to improve efficiency,
```

Comment:
I get that you don't want to load nodes for efficiency reasons, but is this really a desired behavior? If the user checks for non existent groups it will not say they don't exist, it will just think these are virtual. For example:

<pre><code>
In [3]: orm.Group.objects.get_or_create(label= 'a', type_string=orm.GroupTypeString.USER.value )
In [4]: orm.Group.objects.get_or_create(label= 'b/c', type_string=orm.GroupTypeString.USER.value )

In [5]: group_path['a'].is_virtual
Out[5]: False

In [6]: group_path['b'].is_virtual
Out[6]: True

In [7]: group_path['d'].is_virtual
Out[7]: True
</code></pre>

This..."lack of sensitivity"? then propagates to other functions, so for example:

<pre><code>
In [10]: list( group_path['d'].children )
Out[10]: []

In [11]: list( group_path['d/e'].children )
Out[11]: []

In [12]: list( group_path['d/e/f'].children )
Out[12]: []

In [13]: group_path['d/e/f'].path_list
Out[13]: ['d', 'e', 'f']

In [14]: group_path['d/e']['f']
Out[14]: GroupPath('d/e/f', type='user')
</code></pre>

I could find no direct way to check if a group actually exists or use methods in a way that will let you know if they are trying to do stuff in non-existent groups. The user has to do this manually by, for example, checking if each group label  is among its parent's children:

```
In [21]: group_path['b'].path in list( child.path for child in group_path['b'].parent.children )
Out[21]: True

In [22]: group_path['d/e/f'].path in list( child.path for child in group_path['d/e/f'].parent.children )
Out[22]: False
```

I think there would be value in at least wrapping this procedure around a 'group_exists' method, and I would argue that we should consider doing this check (by default?) for (almost?) all of the public methods of the class. Maybe it would require a slightly different implementation that uses private methods/properties if this way turns out to be cyclical (i.e. can't use methods inside the 'check' that also need to 'check' the path themselves), but I would think there is a way to do this without actually loading nodes or doing expensive queries.

Again, I don't remember if this was explicitly discussed during the coding week, I am just very wary of functions that have this "_I'm **technically** doing what you asked for_" kind of "malicious compliance" behavior.

Reply:
> I could find no direct way to check if a group actually exists

Can you not just do `group_path['a'].group is not None`? I could add an `exists` method, but that is all it would do.

> I would argue that we should consider doing this check (by default?) for (almost?) all of the public methods of the class

I disagree. This class is intended to work like `pathlib.Path`, in that there is no explicit checks for group/file existence until you actually need the group. I think would be very costly to be making database calls for every operation.

Reply:
Maybe the children one is a little counter-intuitive, and could have an existence check (without loading the node)

Reply:
> Can you not just do group_path['a'].group is not None? I could add an exists method, but that is all it would do.

Ah, ok, yes that looks shorter. It does require you to "load the group", right? Is this as expensive as loading nodes? I mean, probably doesn't matter for one check but maybe if you are doing some intensive processing the farfetched method could be faster?

> I disagree. This class is intended to work like `pathlib.Path`, in that there is no explicit checks for group/file existence until you actually need the group. I think it would be very costly to be making database calls for every operation.

So, I just checked `pathlib` and I think this is not exactly the case. It is true that it doesn't seem to check when you set the path and it does perform some methods without checking, but it also has an `exist` method and it is not letting me do iterations over an inexistent path:

```
In [1]: from pathlib import Path
In [2]: p = Path('.')
In [3]: p = Path('aloha')
In [4]: p.name
Out[4]: 'aloha'

In [5]: p.is_absolute()
Out[5]: False

In [6]: list( p.iterdir() )
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
<ipython-input-7-ede0ddf8d362> in <module>()
----> 1 list( p.iterdir() )

/usr/lib/python3.6/pathlib.py in iterdir(self)
   1079         if self._closed:
   1080             self._raise_closed()
-> 1081         for name in self._accessor.listdir(self):
   1082             if name in {'.', '..'}:
   1083                 # Yielding a path object for these makes little sense

/usr/lib/python3.6/pathlib.py in wrapped(pathobj, *args)
    385         @functools.wraps(strfunc)
    386         def wrapped(pathobj, *args):
--> 387             return strfunc(str(pathobj), *args)
    388         return staticmethod(wrapped)
    389 

FileNotFoundError: [Errno 2] No such file or directory: 'aloha'
```

It also returns the same error when using the `rmdir` method. I don't think this is true for the `walk` and `delete_group`, no?

Reply:
Yep as I've mentioned in the 2nd comment, an existence check at the start of `def children` should clear this up (and raise an error on `delete_group`). 

Thats the mindset though; where possible mirror `pathlib.Path`. To that end, maybe a specific `GroupNotFoundError` exception class would also be a good idea?

Reply:
Ah, yeah, sorry, I missed the second comment as it got posted after I started writing the reply. Yes, I think it would be important at least for the `children`, `walk` and `delete` methods, and I would explicitly create a method that directly checks for this. Maybe also consider the check for the `__getitem__` and the `group` methods? I'm not sure, personally I would do most of them but perhaps a third opinion would be good here.

And yeah, the specific error sounds good, although again I'm not very experienced in when it is convenient to create new error classes so also open to suggestions.

Reply:
> It does require you to "load the group", right? Is this as expensive as loading nodes?

Yes and loading any database ORM entity is relatively slow, because of the additional logic that django/sqlalchemy does to load it into the session

In d6e5f08, `GroupNotFoundError`/`GroupNotUniqueError` are now raised.

> I think there would be value in at least wrapping this procedure around a 'group_exists' method

You actually missed an important point here; most paths won't have an associated group. For example, if you have paths `a/b/c` and `a/d/e`, I don't want `GroupPath('a').children` failing because it is not directly associated with a group.

To that effect, in 73183a1e5dae5352d71a37d32014a4b08aa15f9b, for anything involving `children` (`__iter__`, `__len__`, `walk`), a `NoGroupsInPathError` exception will be raised if the group itself is not associated with a group, or there are no descendant paths that are contain groups. Personally I think this is added complication to just checking if the path has zero children, but I'm not that bothered.

For `__getitem__`/`__truediv__` this should definitely not raise an existence error though, otherwise this would raise an error `(GroupPath() / 'x' / 'y' / 'z').get_or_create_group()`